### PR TITLE
fix the problem that there may be some obsolete data left in storage which cannot be deleted

### LINKED ISSUE #5659: The page write concurrently with SegmentMerge/Split/MergeDelta may not get removed from the PageStorage
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
TEST_F(SegmentOperationTest, Issue5570Case2)
try
{
    {
        SegmentTestOptions options;
        // a smaller pack rows for logical split
        options.db_settings.dt_segment_stable_pack_rows = 100;
        reloadWithOptions(options);
    }

    writeSegment(DELTA_MERGE_FIRST_SEGMENT_ID, 100);
    writeSegment(DELTA_MERGE_FIRST_SEGMENT_ID, 100);
    flushSegmentCache(DELTA_MERGE_FIRST_SEGMENT_ID);
    mergeSegmentDelta(DELTA_MERGE_FIRST_SEGMENT_ID);

    FailPointHelper::enableFailPoint(FailPoints::force_segment_logical_split);
    auto new_seg_id_opt = splitSegment(DELTA_MERGE_FIRST_SEGMENT_ID);
    ASSERT_TRUE(new_seg_id_opt.has_value());
    auto new_seg_id = new_seg_id_opt.value();

    const auto storage_pool = db_context->getGlobalContext().getGlobalStoragePool();
    for (size_t round = 0; round < 50; ++round)
    {
        LOG_DEBUG(log, "beginSegmentMerge");

        // Start a segment merge and suspend it before applyMerge
        auto sp_seg_merge_apply = SyncPointCtl::enableInScope("before_Segment::applyMerge");
        auto th_seg_merge = std::async([&]() {
            mergeSegment(DELTA_MERGE_FIRST_SEGMENT_ID, new_seg_id, true);
        });
        sp_seg_merge_apply.waitAndPause();
        LOG_DEBUG(log, "pausedBeforeApplyMerge");

        // flushed pack
        writeSegment(new_seg_id, 100);
        // flushSegmentCache(new_seg_id); // do not flush

        // Finish the segment merge
        LOG_DEBUG(log, "continueApplyMerge");
        sp_seg_merge_apply.next();
        th_seg_merge.wait();
        LOG_DEBUG(log, "finishApplyMerge");

        // logical split
        FailPointHelper::enableFailPoint(FailPoints::try_segment_logical_split);
        auto new_seg_id2_opt = splitSegment(DELTA_MERGE_FIRST_SEGMENT_ID);
        ASSERT_TRUE(new_seg_id2_opt.has_value());
        new_seg_id = new_seg_id2_opt.value();

        const auto file_usage = storage_pool->getLogFileUsage();
        LOG_DEBUG(log, "log valid size: {}", file_usage.total_valid_size);
    }
    for (const auto & [seg_id, seg] : segments)
    {
        UNUSED(seg);
        deleteRangeSegment(seg_id);
        flushSegmentCache(seg_id);
        mergeSegmentDelta(seg_id);
    }
    storage_pool->gc();
    const auto file_usage = storage_pool->getLogFileUsage();
    LOG_DEBUG(log, "all removed, file usage: {}", file_usage.total_valid_size); // should be 0
}
CATCH
```

### 2. What did you expect to see? (Required)

After the test run, we call delete range and delta merge for all segments, the valid size of the log PageStorage should be zero.

### 3. What did you see instead (Required)

The valid size of log PageStorage is not zero

The root cause is in `applyMergeDelta`/`applyMerge`/`applySplit`, we create ref page for all pages that is not exists in the snapshot, but only remove the original page that is flushed into disk.

image

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close #5659 

Problem Summary: When write to segment happens concurrently with segment split/merge/merge delta, there may be some data left in PageStorage which cannot be deleted. Because we don't clean potential `ColumnFilePersisted` in `MemTableSet` before abandoning the segment.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix the problem that there may be some obsolete data left in storage which cannot be deleted
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp [hunk 1]
```diff
@@ -83,6 +93,188 @@ try
...
+        // Start a segment merge and suspend it before applyMerge
+        auto sp_seg_merge_delta_apply = SyncPointCtl::enableInScope("before_Segment::applyMergeDelta");
+        auto th_seg_merge_delta = std::async([&]() {
+            mergeSegmentDelta(DELTA_MERGE_FIRST_SEGMENT_ID, /* check_rows */ false);
+        });
+        sp_seg_merge_delta_apply.waitAndPause();
+
+        LOG_DEBUG(log, "pausedBeforeApplyMergeDelta");
+
+        // flushed pack
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp:117]:
```suggestion
        // non-flushed pack
```

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp [hunk 2]
```diff
@@ -83,6 +93,188 @@ try
...
+        auto th_seg_split = std::async([&]() {
+            auto new_seg_id_opt = splitSegment(DELTA_MERGE_FIRST_SEGMENT_ID, /* check_rows */ false);
+            ASSERT_TRUE(new_seg_id_opt.has_value());
+            new_seg_id = new_seg_id_opt.value();
+        });
+        sp_seg_split_apply.waitAndPause();
+
+        LOG_DEBUG(log, "pausedBeforeApplySplit");
+
+        // flushed pack
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp:178]: ditto

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp [hunk 3]
```diff
@@ -83,6 +93,188 @@ try
...
+        // Start a segment merge and suspend it before applyMerge
+        auto sp_seg_merge_apply = SyncPointCtl::enableInScope("before_Segment::applyMerge");
+        auto th_seg_merge = std::async([&]() {
+            mergeSegment(DELTA_MERGE_FIRST_SEGMENT_ID, new_seg_id, /* check_rows */ false);
+        });
+        sp_seg_merge_apply.waitAndPause();
+
+        LOG_DEBUG(log, "pausedBeforeApplyMerge");
+
+        // flushed pack
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp:241]: ditto

**Comment by JaySon-Huang** [comment]
The unit test failed with a similar problem as https://github.com/pingcap/tiflash/issues/5570. Pending for https://github.com/pingcap/tiflash/pull/5612 to be merged first in the master branch

File: dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp [hunk 4]
```diff
@@ -97,6 +97,7 @@ size_t DeltaValueSpace::getValidCacheRows() const
 void DeltaValueSpace::recordRemoveColumnFilesPages(WriteBatches & wbs) const
 {
     persisted_file_set->recordRemoveColumnFilesPages(wbs);
+    mem_table_set->recordRemoveColumnFilesPages(wbs);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp:100]:
```suggestion
    // there could be some persisted column files in the `mem_table_set` which should be removed.
    mem_table_set->recordRemoveColumnFilesPages(wbs);
```

**Review by JaySon-Huang — COMMENTED** [review]
Rest LGTM

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp (hunk 1 above)
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp]: Fixed.

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp (hunk 2 above)
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp]: Fixed.

File: dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp (hunk 3 above)
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/tests/gtest_segment.cpp]: Fixed.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**